sphinx/environment/__init__.py

```diff
@@ -416,37 +416,40 @@ def find_files(self, config, builder):
-            else:
-                logger.warning(__("document not readable. Ignored."), location=docname)
-
-        # Current implementation is applying translated messages in the reading
-        # phase.Therefore, in order to apply the updated message catalog, it is
-        # necessary to re-process from the reading phase. Here, if dependency
-        # is set for the doc source and the mo file, it is processed again from
-        # the reading phase when mo is updated. In the future, we would like to
-        # move i18n process into the writing phase, and remove these lines.
-        if builder.use_message_catalog:
-            # add catalog mo file dependency
-            for docname in self.found_docs:
-                catalog_files = find_catalog_files(
-                    docname,
-                    self.srcdir,
-                    self.config.locale_dirs,
-                    self.config.language,
-                    self.config.gettext_compact)
-                for filename in catalog_files:
-                    self.dependencies[docname].add(filename)
+        try:
+            matchers = compile_matchers(
+                config.exclude_patterns[:] +
+                config.templates_path +
+                builder.get_asset_paths() +
+                ['**/_sources', '.#*', '**/.#*', '*.lproj/**']
+            )
+            self.found_docs = set()
+            for docname in get_matching_docs(self.srcdir, config.source_suffix,  # type: ignore
+                                             exclude_matchers=matchers):
+                if os.access(self.doc2path(docname), os.R_OK):
+                    self.found_docs.add(docname)
+                else:
+                    logger.warning(__("document not readable. Ignored."), location=docname)
+
+            # Current implementation is applying translated messages in the reading
+            # phase.Therefore, in order to apply the updated message catalog, it is
+            # necessary to re-process from the reading phase. Here, if dependency
+            # is set for the doc source and the mo file, it is processed again from
+            # the reading phase when mo is updated. In the future, we would like to
+            # move i18n process into the writing phase, and remove these lines.
+            if builder.use_message_catalog:
+                # add catalog mo file dependency
+                for docname in self.found_docs:
+                    catalog_files = find_catalog_files(
+                        docname,
+                        self.srcdir,
+                        self.config.locale_dirs,
+                        self.config.language,
+                        self.config.gettext_compact)
+                    for filename in catalog_files:
+                        self.dependencies[docname].add(filename)
+        except OSError as exc:
+            raise SphinxError(__('Failed to scan documents in %s: %r') % (self.srcdir, exc))
```

Comment:
@shimizukawa I'd like to your opinion to this patch.

1. Is `SphinxError` appropriate for this case? Should I add a new error class like `BuildError` or `DocumentError`?
2. Which module is better to handle such exceptions? `env`, `builder` or `app`?

Reply:
1. +1 for `BuildError` and/or `DocumentError` instead of direct usage of `SphinxError`. 
2. I think that these primitive exceptions should be caught by the direct use code. Instead, I think it's better to catch SphinxError and its descendant exceptions in one place and deal with it.

Reply:
done